rdm/0007-access-control.md

```diff
@@ -0,0 +1,222 @@
+- As a user, I want to access my data even after I have left my university.
+- As an administrator, I want to open my repositories for other institutions and their IDMS (?? Institutional Data Management System).
+
+## Terminology
+
+- **Resources**: The term resource in the system represents anything you want to protect. Resources can both be at the object-level (e.g. access to a record) or of more general nature (e.g access to the administration interface).
+Actions: Resources can be protected by requiring the ability to perform one or more actions. 
+- **Subjects**: Access to resources can be granted to *subjects*, which include users, roles and system roles. System roles are roles that are programmatically assigned to users. For example:
+    - Authenticated user: any logged in user.
+    - Any user: both authenticated and anonymous users.
+    - Campus user: Assigned to users based on their IP address being in a specific range.
+
+
+## Detailed design
+
+Invenio implements a Role-Based Access Control system (RBAC), that can be used to protect and grant access to resources.
+
+### Records as resources
+
+The most complex access rights use cases in Invenio RDM are related to records. Note that the following access control can be different between records and deposit records. In addition, this access control model is not tight to Invenio RDM, so it can be applied to any data model in Invenio.
+
+**Record actions**
+
+All records have the following basic CRUD actions:
+- Create: Ability to create new records.
+- Read: Ability to read a record (including during searches).
+- Update: Ability to update a record metadata.
+- Delete: Ability to delete a record.
+- List: Ability to perform searches (note, for a record to appear in the search result it is subject to the above mentioned read action).
+
+Records linked with files (possible for all records) have the additional following actions.
+- Create files: Ability to create files associated with a record. 
+- Read files: Ability to read files associated with a record.
+- Update files: Ability to update files associated with a record.
+- Delete files: Ability to delete files associated with a record.
+- Administrate files: Ability to perform all possible actions on files associated with a record, including reading previous versions of files.
+
+Deposit records have the additional following actions:
+- Publish: Ability to publish a new record.
+- Edit: Ability to edit an already published record. This differs from the record update action.
+- New version: Ability to create a new version of an existing record.
+
+It is possible to extend this model with more actions for each data model as needed.
+
+**Requiring permissions**
+
+A record must store permissions inside the metadata. This can be as simple as a list of owners, or very complex based on custom rules (e.g. books can only be curated by books curators). This rules are translated into `generators`.
```

Comment:
I think the distinction should be made between system-wide policies and record-specific policies i.e. implicit versus explicit permissions.

```suggestion
A record must store permissions inside the metadata. This can be as implicit as relying on a list of owners, or very explicit and detailed based on encoded rules (e.g. books can only be curated by books curators encoded in the record). These rules are translated into `generators`.

Implicit access rules are typically set at the instance-level, while explicit access rules are typically set on a resource by resource basis.
```

Obviously, it's instance-level policies that enable access rules at the resource level...

Reply:
I agree on making the distinction. However, isn't the list of owners **explicit** and the ones encoded according to business rules **implicit**?

Reply:
I think the *list* of owners is explicit, but the *associated* permissions are implicit. If I recall correctly, some institutions only want mediated deposits: in this case, being an owner is explicit on the record, but that doesn't give the owner edit permissions because only creators/curators have edit permissions at this institution.

In any case, the permission policy is the ultimate decider: even if there is an explicit encoding in a record, it is the defined generators that interpret/enforce it.

Perhaps the text was describing the effort involved in / look of encoded permissions in metadata? In which case I agree, it can vary from just listing owners to encoding complex rules.